gamedata/ddb.py

```diff
@@ -0,0 +1,417 @@
+import json
+import logging
+import time
+
+import aioboto3
+import aiohttp
+import jwt
+from boto3.dynamodb.conditions import Key
+
+from cogsmisc.stats import Stats
+# auth service
+from utils.config import DDB_AUTH_AUDIENCE as AUDIENCE, \
+    DDB_AUTH_EXPIRY_SECONDS as EXPIRY_SECONDS, \
+    DDB_AUTH_ISSUER as ISSUER, \
+    DDB_AUTH_SECRET as MY_SECRET, \
+    DDB_WATERDEEP_SECRET as WATERDEEP_SECRET, \
+    DDB_AUTH_SERVICE_URL as AUTH_BASE_URL
+# dynamo
+# env: AWS_ACCESS_KEY_ID
+# env: AWS_SECRET_ACCESS_KEY
+from utils.config import DYNAMO_ENTITY_TABLE, DYNAMO_REGION, DYNAMO_USER_TABLE
+
+AUTH_DISCORD = f"{AUTH_BASE_URL}/v1/discord-token"
+
+# cache
+USER_ENTITLEMENT_TTL = 1 * 60
+ENTITY_ENTITLEMENT_TTL = 15 * 60
+
+log = logging.getLogger(__name__)
+
+
+class BeyondClientBase:  # for development - assumes no entitlements
+    async def get_accessible_entities(self, ctx, user_id, entity_type):
+        return None
+
+    async def get_ddb_user(self, ctx, user_id):
+        return None
+
+    async def close(self):
+        pass
+
+
+class BeyondClient(BeyondClientBase):
+    """
+    Client to interface with DDB's Auth Service and Entitlements tables in DynamoDB.
+    Asyncio-compatible.
+
+    Most methods are private since local dev environments cannot connect to the DDB stack, and
+    public methods should return the most conservative permissions (i.e. user owns nothing in db)
+    possible without making external connections in this scenario.
+    """
+
+    def __init__(self, loop):
+        self.http = None
+        self.dynamo = None
+        self.ddb_user_table = None
+        self.ddb_entity_table = None
+        loop.run_until_complete(self._initialize())
+
+    async def _initialize(self):
+        """Initialize our async resources: aiohttp, aioboto3"""
+        self.http = aiohttp.ClientSession()
+        self.dynamo = aioboto3.resource('dynamodb', region_name=DYNAMO_REGION)
+        self.ddb_user_table = self.dynamo.Table(DYNAMO_USER_TABLE)
+        self.ddb_entity_table = self.dynamo.Table(DYNAMO_ENTITY_TABLE)
+        log.info("DDB client initialized")
+
+    async def get_accessible_entities(self, ctx, user_id, entity_type):
```

Comment:
Main Entitlements method: returns the set of entity IDs of the given entity type the given user can see, or None if there is no DDB link.

Hits the Auth Service to get the DDB user, and the entitlements tables for the license->entity mapping.